javascript/ql/lib/semmle/javascript/NPM.qll

```diff
@@ -12,8 +12,23 @@ class PackageJson extends JsonObject {
     this.isTopLevel()
   }
 
-  /** Gets the name of this package. */
-  string getPackageName() { result = this.getPropStringValue("name") }
+  /**
+   * Gets the name of this package.
+   * If the package is located under the package `pkg1` and its relative path is `foo/bar`, it can be `pkg1/foo/bar`.
+   */
+  string getPackageName() {
+    result = this.getPropStringValue("name")
+    or
+    exists(PackageJson parentPkg, Container currentDir, Container parentDir |
+      currentDir = this.getJsonFile().getParentContainer() and
+      parentDir = parentPkg.getJsonFile().getParentContainer() and
+      parentDir.getParentContainer+().getBaseName() = "node_modules" and
```

Comment:
```suggestion
      parentDir.getParentContainer().getBaseName() = "node_modules" and
```

I'm not sure I like that transitive closure.  
Your test only covers that the `parentDir´ is directly contained within `node_modules`, and I'm not sure what happens if `parentDir` is not directly inside `node_modules`.

Reply:
> Do you think it is better to rewrite this code to the following one?

Yes, that could work. 
And in the `.getParentContainer().getParentContainer()` case you can check that the packagename of the `parentPkg` contains `"/"`.

Reply:
Hi, I fixed the code.
Thanks for your review!